svgdigitizer/svgplot.py

```diff
@@ -920,120 +1116,264 @@ def sample_path(self, path, sampling_interval, endpoints="include"):
+            ([], 1.0)
+
+        When including the endpoints, `sample_from_x_length` has no effect since
+        we always start and end sampling at the end points of the segment::
+
+            >>> SVGPlot._sample_segment(segment, .25, sample_from_x_length=2)
+            ([0, 0.25, 0.5, 0.75, 1], 0.0)
+
+        ::
+
+            >>> SVGPlot._sample_segment(segment, .75, sample_from_x_length=2)
+            ([0, 0.75, 1], 0.0)
+
+        """
+        if sample_from_x_length < 0:
+            raise ValueError(f"Cannot sample at negative length {sample_from_x_length}")
+
+        sample_at = []
+
+        # The path length on the x-axis at which we plan to sample (in the range [0, length of the path segment]):
+        x_length_target = sample_from_x_length
+
+        if endpoints == "include":
+            # Include the initial point of the path segment.
+            sample_at.append(0)
+
+            # Continue sampling from the start of the path segment.
+            x_length_target = sampling_interval
+
+        import numpy
+
+        x = numpy.poly1d(numpy.real(segment.poly()))
+
+        # At the extrema of the projection to the x-axis, the curve changes direction.
+        extrema = list(
+            sorted(
+                root.real
+                for root in numpy.polyder(x).roots
+                if abs(root.imag) < cls._EPSILON and 0 < root.real < 1
+            )
+        )
+
+        # Eventually this will equal the total length of this path segment.
+        segment_length = 0
+
+        # Sample in the range [tmin, tmax] = [time at which the curve changed its behavior last, time at which the curve changes its behavior next]
+        for tmin, tmax in zip([0] + extrema, extrema + [1]):
+            snippet_length = abs(x(tmax) - x(tmin))
+
+            sample_snippet_at, x_length_target = cls._sample_snippet(
+                segment=segment,
+                t_range=(tmin, tmax),
+                sampling_interval=sampling_interval,
+                sample_from_x_length=x_length_target,
+                x_length_range=(segment_length, segment_length + snippet_length),
+            )
+            sample_at.extend(sample_snippet_at)
+
+            segment_length += snippet_length
+
+        # We have left the path segment.
+        if endpoints == "include":
+            # Include the final point of this path segment.
+            sample_at.append(1)
+
+        assert sample_at == sorted(sample_at), f"Samples are out of order {sample_at}"
+
+        if endpoints == "include":
+            # Do not sample points that are just a numerical-error away from the end points.
+            assert (
+                sample_at[1] > cls._EPSILON
+            ), f"First real sampling point should be quite a bit away from t=0 but it is only at {sample_at[1]}"
+            if 1 - sample_at[-2] < cls._EPSILON:
+                sample_at = sample_at[:-2] + [sample_at[-1]]
+
+            x_length_target = segment_length
```

Comment:
This line was missing before and might cause some slight changes in some outputs.